File: iwa-client/src/stun-client.js
/**
 * STUN Client for IWA Direct Sockets
 *
 * Implements RFC 5389 STUN Binding Request/Response
 * Uses dual-query to detect symmetric NAT
 */

// STUN message types
const BINDING_REQUEST = 0x0001;
const BINDING_RESPONSE = 0x0101;

// STUN attribute types
const MAPPED_ADDRESS = 0x0001;
const XOR_MAPPED_ADDRESS = 0x0020;

// STUN magic cookie (RFC 5389)
const MAGIC_COOKIE = 0x2112a442;

// Default STUN servers for NAT detection
const STUN_SERVERS = [
  { host: "stun.cloudflare.com", port: 3478 },
  { host: "stun.l.google.com", port: 19302 },
];

/**
 * Generate random transaction ID (12 bytes)
 */
function generateTransactionId() {
  const id = new Uint8Array(12);
  crypto.getRandomValues(id);
  return id;
}

/**
 * Build STUN Binding Request (20 bytes header, no attributes)
 */
function buildBindingRequest(transactionId) {
  const buffer = new ArrayBuffer(20);
  const view = new DataView(buffer);

  // Message Type: Binding Request (0x0001)
  view.setUint16(0, BINDING_REQUEST, false);

  // Message Length: 0 (no attributes)
  view.setUint16(2, 0, false);

  // Magic Cookie
  view.setUint32(4, MAGIC_COOKIE, false);

  // Transaction ID (12 bytes)
  const arr = new Uint8Array(buffer);
  arr.set(transactionId, 8);

  return new Uint8Array(buffer);
}

/**
 * Parse STUN Binding Response, extract XOR-MAPPED-ADDRESS
 * Returns { ip: string, port: number } or null
 */
function parseBindingResponse(data, transactionId) {
  if (data.byteLength < 20) return null;

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);

  // Check message type
  const msgType = view.getUint16(0, false);
  if (msgType !== BINDING_RESPONSE) return null;

  // Check magic cookie
  const cookie = view.getUint32(4, false);
  if (cookie !== MAGIC_COOKIE) return null;

  // Check transaction ID
  for (let i = 0; i < 12; i++) {
    if (data[8 + i] !== transactionId[i]) return null;
  }

  // Parse attributes
  const msgLength = view.getUint16(2, false);
  let offset = 20;
  const end = 20 + msgLength;

  while (offset + 4 <= end) {
    const attrType = view.getUint16(offset, false);
    const attrLength = view.getUint16(offset + 2, false);
    const attrStart = offset + 4;

    if (attrType === XOR_MAPPED_ADDRESS && attrLength >= 8) {
      // Family (1 byte padding + 1 byte family)
      const family = view.getUint8(attrStart + 1);

      if (family === 0x01) {
        // IPv4
        // XOR'd port (with upper 16 bits of magic cookie)
        const xorPort = view.getUint16(attrStart + 2, false);
        const port = xorPort ^ (MAGIC_COOKIE >>> 16);

        // XOR'd address (with magic cookie)
        const xorAddr = view.getUint32(attrStart + 4, false);
        const addr = xorAddr ^ MAGIC_COOKIE;

        const ip = [
          (addr >>> 24) & 0xff,
          (addr >>> 16) & 0xff,
          (addr >>> 8) & 0xff,
          addr & 0xff,
        ].join(".");

        return { ip, port };
      }
      // IPv6 not supported for now
    } else if (attrType === MAPPED_ADDRESS && attrLength >= 8) {
      // Fallback to non-XOR mapped address (older servers)
      const family = view.getUint8(attrStart + 1);

      if (family === 0x01) {
        const port = view.getUint16(attrStart + 2, false);
        const addr = view.getUint32(attrStart + 4, false);

        const ip = [
          (addr >>> 24) & 0xff,
          (addr >>> 16) & 0xff,
          (addr >>> 8) & 0xff,
          addr & 0xff,
        ].join(".");

        return { ip, port };
      }
    }

    // Move to next attribute (aligned to 4 bytes)
    offset = attrStart + Math.ceil(attrLength / 4) * 4;
  }

  return null;
}

/**
 * Query a single STUN server
 * @param {ReadableStream} readable - Socket readable stream
 * @param {WritableStreamDefaultWriter} writer - Writer for the socket
 * @param {string} host - STUN server hostname
 * @param {number} port - STUN server port
 * @param {number} timeout - Timeout in ms
 * @returns {Promise<{ip: string, port: number} | null>}
 */
async function queryStunServer(readable, writer, host, port, timeout = 3000) {
  const transactionId = generateTransactionId();
  const request = buildBindingRequest(transactionId);

  // Resolve hostname using DNS-over-HTTPS
  const { addresses } = await fetch(
    `https://dns.google/resolve?name=${host}&type=A`,
  )
    .then((r) => r.json())
    .then((data) => ({
      addresses:
        data.Answer?.filter((a) => a.type === 1).map((a) => a.data) || [],
    }));

  if (addresses.length === 0) {
    throw new Error(`Failed to resolve ${host}`);
  }

  const serverIp = addresses[0];

  // Send STUN request
  await writer.write({
    data: request,
    remoteAddress: serverIp,
    remotePort: port,
  });

  // Read responses with timeout
  const startTime = Date.now();
  const reader = readable.getReader();

  try {
    while (Date.now() - startTime < timeout) {
      const readPromise = reader.read();
      const timeoutPromise = new Promise((resolve) =>
        setTimeout(
          () => resolve({ timeout: true }),
          Math.max(100, timeout - (Date.now() - startTime)),
        ),
      );

      const result = await Promise.race([readPromise, timeoutPromise]);

      if (result.timeout) {
        // Retry once
        await writer.write({
          data: request,
          remoteAddress: serverIp,
          remotePort: port,
        });
        continue;
      }

      if (result.done) break;

      const { data, remoteAddress, remotePort } = result.value;

      // Only process responses from the server we queried
      if (remoteAddress !== serverIp || remotePort !== port) continue;

      const parsed = parseBindingResponse(new Uint8Array(data), transactionId);
      if (parsed) {
        return parsed;
      }
    }
  } finally {
    reader.releaseLock();
  }

  return null;
}

/**
 * Discover public endpoint with NAT type detection
 *
 * Queries two different STUN servers from the same socket.
 * If both return different ports, it's symmetric NAT.
 *
 * @param {ReadableStream} readable - Socket readable stream
 * @param {WritableStreamDefaultWriter} writer - Socket writer
 * @returns {Promise<{publicIp: string, publicPort: number, isSymmetric: boolean}>}
 */
export async function discoverEndpoint(readable, writer) {
  // Query STUN servers sequentially (same socket, can't read in parallel)
  const result1 = await queryStunServer(
    readable,
    writer,
    STUN_SERVERS[0].host,
    STUN_SERVERS[0].port,
  );
  const result2 = await queryStunServer(
    readable,
    writer,
    STUN_SERVERS[1].host,
    STUN_SERVERS[1].port,
  );

  if (!result1 && !result2) {
    throw new Error("Failed to contact any STUN server");
  }

  // Use first successful result as public endpoint
  const primary = result1 || result2;

  // Check for symmetric NAT (different ports for different destinations)
  let isSymmetric = false;
  if (result1 && result2) {
    isSymmetric = result1.port !== result2.port;
  }

  return {
    publicIp: primary.ip,
    publicPort: primary.port,
    isSymmetric,
  };
}

/**
 * Simple single-server STUN query (for cases where NAT detection not needed)
 * @param {ReadableStream} readable - Socket readable stream
 * @param {WritableStreamDefaultWriter} writer - Socket writer
 * @returns {Promise<{publicIp: string, publicPort: number}>}
 */
export async function getPublicEndpoint(readable, writer) {
  const result = await queryStunServer(
    readable,
    writer,
    STUN_SERVERS[0].host,
    STUN_SERVERS[0].port,
  );

  if (!result) {
    throw new Error("STUN query failed");
  }

  return {
    publicIp: result.ip,
    publicPort: result.port,
  };
}

File: pi-scripts/stun_client.py
"""
STUN Client for Pi

Implements RFC 5389 STUN Binding Request/Response
Uses dual-query to detect symmetric NAT

Usage:
    from stun_client import discover_endpoint, get_public_endpoint
    
    # Full NAT detection
    endpoint = await discover_endpoint(sock)
    print(f"Public: {endpoint['ip']}:{endpoint['port']}, Symmetric: {endpoint['is_symmetric']}")
    
    # Simple single query
    endpoint = await get_public_endpoint(sock)
"""

import asyncio
import socket
import struct
import os
import logging

logger = logging.getLogger(__name__)

# STUN message types
BINDING_REQUEST = 0x0001
BINDING_RESPONSE = 0x0101

# STUN attribute types
MAPPED_ADDRESS = 0x0001
XOR_MAPPED_ADDRESS = 0x0020

# STUN magic cookie (RFC 5389)
MAGIC_COOKIE = 0x2112A442

# Default STUN servers for NAT detection
STUN_SERVERS = [
    ("stun.cloudflare.com", 3478),
    ("stun.l.google.com", 19302),
]


def generate_transaction_id() -> bytes:
    """Generate random 12-byte transaction ID."""
    return os.urandom(12)


def build_binding_request(transaction_id: bytes) -> bytes:
    """Build STUN Binding Request (20 bytes header, no attributes)."""
    return struct.pack(
        ">HHI",
        BINDING_REQUEST,  # Message type
        0,                # Message length (no attributes)
        MAGIC_COOKIE,     # Magic cookie
    ) + transaction_id


def parse_binding_response(data: bytes, transaction_id: bytes) -> dict | None:
    """
    Parse STUN Binding Response, extract XOR-MAPPED-ADDRESS.
    Returns {'ip': str, 'port': int} or None.
    """
    if len(data) < 20:
        return None

    # Parse header
    msg_type, msg_length, cookie = struct.unpack(">HHI", data[:8])

    if msg_type != BINDING_RESPONSE:
        return None
    if cookie != MAGIC_COOKIE:
        return None
    if data[8:20] != transaction_id:
        return None

    # Parse attributes
    offset = 20
    end = 20 + msg_length

    while offset + 4 <= end:
        attr_type, attr_length = struct.unpack(">HH", data[offset:offset + 4])
        attr_start = offset + 4

        if attr_type == XOR_MAPPED_ADDRESS and attr_length >= 8:
            # Reserved (1 byte) + Family (1 byte) + Port (2 bytes) + Address (4 bytes)
            _, family, xor_port, xor_addr = struct.unpack(
                ">BBHI", data[attr_start:attr_start + 8]
            )

            if family == 0x01:  # IPv4
                port = xor_port ^ (MAGIC_COOKIE >> 16)
                addr = xor_addr ^ MAGIC_COOKIE

                ip = socket.inet_ntoa(struct.pack(">I", addr))
                return {"ip": ip, "port": port}

        elif attr_type == MAPPED_ADDRESS and attr_length >= 8:
            # Fallback to non-XOR mapped address
            _, family, port, addr = struct.unpack(
                ">BBHI", data[attr_start:attr_start + 8]
            )

            if family == 0x01:  # IPv4
                ip = socket.inet_ntoa(struct.pack(">I", addr))
                return {"ip": ip, "port": port}

        # Move to next attribute (aligned to 4 bytes)
        offset = attr_start + ((attr_length + 3) // 4) * 4

    return None


async def query_stun_server(
    sock: socket.socket,
    host: str,
    port: int,
    timeout: float = 3.0
) -> dict | None:
    """
    Query a single STUN server.
    
    Args:
        sock: Bound UDP socket
        host: STUN server hostname
        port: STUN server port
        timeout: Timeout in seconds
        
    Returns:
        {'ip': str, 'port': int} or None
    """
    loop = asyncio.get_event_loop()
    transaction_id = generate_transaction_id()
    request = build_binding_request(transaction_id)

    # Resolve hostname
    try:
        infos = await loop.getaddrinfo(host, port, family=socket.AF_INET, type=socket.SOCK_DGRAM)
        if not infos:
            logger.warning(f"Failed to resolve {host}")
            return None
        server_addr = infos[0][4]  # (ip, port)
    except Exception as e:
        logger.warning(f"DNS resolution failed for {host}: {e}")
        return None

    # Send request
    try:
        sock.sendto(request, server_addr)
    except Exception as e:
        logger.warning(f"Failed to send STUN request to {host}: {e}")
        return None

    # Wait for response with retries
    sock.setblocking(False)
    start_time = asyncio.get_event_loop().time()

    while True:
        elapsed = asyncio.get_event_loop().time() - start_time
        if elapsed >= timeout:
            return None

        remaining = timeout - elapsed

        try:
            # Use asyncio to wait for socket to be readable
            await asyncio.wait_for(
                loop.sock_recv(sock, 1024),
                timeout=min(remaining, 0.5)
            )
        except asyncio.TimeoutError:
            # Retry
            try:
                sock.sendto(request, server_addr)
            except:
                pass
            continue
        except Exception as e:
            logger.warning(f"Error receiving STUN response: {e}")
            continue

    return None


async def query_stun_server_simple(
    sock: socket.socket,
    host: str,
    port: int,
    timeout: float = 3.0
) -> dict | None:
    """
    Query a single STUN server (simpler implementation using threads).
    
    Args:
        sock: Bound UDP socket
        host: STUN server hostname  
        port: STUN server port
        timeout: Timeout in seconds
        
    Returns:
        {'ip': str, 'port': int} or None
    """
    loop = asyncio.get_event_loop()
    transaction_id = generate_transaction_id()
    request = build_binding_request(transaction_id)

    # Resolve hostname
    try:
        infos = await loop.getaddrinfo(host, port, family=socket.AF_INET, type=socket.SOCK_DGRAM)
        if not infos:
            logger.warning(f"Failed to resolve {host}")
            return None
        server_addr = infos[0][4]  # (ip, port)
    except Exception as e:
        logger.warning(f"DNS resolution failed for {host}: {e}")
        return None

    def do_stun():
        """Blocking STUN query."""
        sock.settimeout(timeout)
        retries = 3
        
        try:
            for attempt in range(retries):
                try:
                    sock.sendto(request, server_addr)
                    data, addr = sock.recvfrom(1024)
                    
                    result = parse_binding_response(data, transaction_id)
                    if result:
                        return result
                        
                except socket.timeout:
                    continue
                except Exception as e:
                    logger.warning(f"STUN query attempt {attempt + 1} failed: {e}")
                    continue
            
            return None
        finally:
            # Restore non-blocking mode for asyncio compatibility
            sock.setblocking(False)

    # Run blocking operation in thread pool
    return await loop.run_in_executor(None, do_stun)


async def discover_endpoint(sock: socket.socket) -> dict:
    """
    Discover public endpoint with NAT type detection.
    
    Queries two different STUN servers from the same socket.
    If both return different ports, it's symmetric NAT.
    
    Args:
        sock: Bound UDP socket
        
    Returns:
        {
            'ip': str,
            'port': int,
            'is_symmetric': bool
        }
        
    Raises:
        RuntimeError: If all STUN servers fail
    """
    # Query both servers (not in parallel - same socket)
    result1 = await query_stun_server_simple(
        sock, STUN_SERVERS[0][0], STUN_SERVERS[0][1]
    )
    result2 = await query_stun_server_simple(
        sock, STUN_SERVERS[1][0], STUN_SERVERS[1][1]
    )

    if not result1 and not result2:
        raise RuntimeError("Failed to contact any STUN server")

    # Use first successful result as public endpoint
    primary = result1 or result2

    # Check for symmetric NAT (different ports for different destinations)
    is_symmetric = False
    if result1 and result2:
        is_symmetric = result1["port"] != result2["port"]
        if is_symmetric:
            logger.warning(
                f"Symmetric NAT detected: port {result1['port']} vs {result2['port']}"
            )

    logger.info(
        f"STUN discovery: {primary['ip']}:{primary['port']} "
        f"(symmetric={is_symmetric})"
    )

    return {
        "ip": primary["ip"],
        "port": primary["port"],
        "is_symmetric": is_symmetric,
    }


async def get_public_endpoint(sock: socket.socket) -> dict:
    """
    Simple single-server STUN query (for cases where NAT detection not needed).
    
    Args:
        sock: Bound UDP socket
        
    Returns:
        {'ip': str, 'port': int}
        
    Raises:
        RuntimeError: If STUN query fails
    """
    result = await query_stun_server_simple(
        sock, STUN_SERVERS[0][0], STUN_SERVERS[0][1]
    )

    if not result:
        raise RuntimeError("STUN query failed")

    logger.info(f"Public endpoint: {result['ip']}:{result['port']}")
    return result


# For testing
if __name__ == "__main__":
    async def main():
        logging.basicConfig(level=logging.INFO)
        
        # Create and bind socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.bind(("0.0.0.0", 0))
        local_port = sock.getsockname()[1]
        print(f"Local port: {local_port}")
        
        try:
            endpoint = await discover_endpoint(sock)
            print(f"Public IP: {endpoint['ip']}")
            print(f"Public Port: {endpoint['port']}")
            print(f"Symmetric NAT: {endpoint['is_symmetric']}")
        finally:
            sock.close()

    asyncio.run(main())
